### ISSUE TITLE
[Spell/Mechanic] Refreshing damage and healing over time (de)buffs is handled incorrectly

### ISSUE THREAD

**Issue by questbugreporter** [WoWManiaUK/Blackwing-Lair#4121]
**Links:**
https://wow.gamepedia.com/Damage_over_Time
https://web.archive.org/web/20111029194001/http://elitistjerks.com/f73/t110353-balance_cataclysm_4_2_a/

**What is happening:**

If you recast a HoT or DoT effect on a target, the duration isn't extended unless there is only one tick remaining. Some spells don't even seem to follow the one tick remaining rule.

**What should happen:**

From the gamepedia article:

>Also as of patch 4.0.1, DoTs no longer can be 'clipped'. Previously, if a DoT was refreshed before it had expired, the tick timer was reset, causing a delay of up to double the normal time between ticks between the last tick that occurred and the first tick of the refreshed DoT. Now, the tick timer is constant. If a DoT is refreshed before it expires, the duration of the DoT will be extended by up to the full normal duration of the DoT, and ticks will continue to occur at the prior interval. Between patches 4.0.1 and 6.0.1, this functionality only allowed DoTs to be "clipped" by a single tick before duration was wasted, requiring DoTs to be refreshed between the second-to-last and the last tick to maintain uninterrupted DoT ticks while optimizing duration gain.

From the elitist jerks archive:

>Cataclysm changed the way DoT refreshing works--when you refresh a currently ticking DoT, you no longer waste a partial tick. Significantly, if you refresh a DoT when it has only one tick remaining, there is no loss at all.

Basically when you refresh a dot or hot effect the new duration should be equal to the base duration plus whatever time was left until the next tick of the dot. Recasting a dot/hot should not delay its damage/healing at all.
